File: C/61. Rotate List.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* rotateRight(struct ListNode* head, int k) {
    int len, i;
    struct ListNode *cur;
    struct ListNode *slow, *fast;

    /* Find actual operation steps. */
    cur = head;
    for (len = 0; cur != NULL; ++len) {
        cur = cur->next;
    }
    if (len == 0) {
        return NULL;
    } else {
        k %= len;
        /* Rotate list according to k. */
        slow = fast = head;
        for (i = 0; i < k; ++i) {
            fast = fast->next;
        }
        while (fast->next != NULL) {
            slow = slow->next;
            fast = fast->next;
        }
        fast->next = head;
        head = slow->next;
        slow->next = NULL;
        return head;
    }
}

File: Python3/61. Rotate List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if not head:
            return None
        else:
            # Determine actual k here if k is lager than size of list
            size = self.getSize(head)
            k %= size

            if k == 0:
                return head
            else:
                # Find k-th node from the end of list
                fast = slow = head
                for _ in range(k):
                    fast = fast.next
                while fast:
                    sprev, fprev = slow, fast
                    slow, fast = slow.next, fast.next

                # Rotate list now
                sprev.next, fprev.next = None, head
                return slow

    def getSize(self, head: ListNode) -> int:
        count = 0
        while head:
            count += 1
            head = head.next
        return count